src/com/twitter/intellij/pants/settings/PantsProjectSettingsControl.java

```diff
@@ -112,65 +120,84 @@ protected boolean isExtraSettingModified() {
-      myTargetSpecsBox.setEmptyText(StatusText.DEFAULT_EMPTY_TEXT);
-      myLibsWithSourcesCheckBox.setEnabled(true);
-
-      ProgressManager.getInstance().run(new Task.Modal(getProject(), PantsBundle.message("pants.getting.target.list"), false) {
-        @Override
-        public void run(@NotNull ProgressIndicator indicator) {
-          try {
-            final Collection<String> targets = PantsUtil.listAllTargets(projectPath);
-            UIUtil.invokeLaterIfNeeded(() -> {
+    ProjectPathFileType pathFileType = determinePathKind(file);
+    switch (pathFileType) {
+      case nonExistent:
+      case nonPantsFile:
+        myTargetSpecsBox.setEnabled(true);
+        errors.add(String.format("Pants project not found given project path: %s", projectPath));
+        break;
+
+      case recursiveDirectory:
+        myTargetSpecsBox.setEnabled(false);
+        Optional<String> relativeProjectPath = PantsUtil.getRelativeProjectPath(file.getPath());
+
+        if (relativeProjectPath.isPresent()) {
+          String spec = relativeProjectPath.get() + "/::";
+
+          myTargetSpecsBox.setEmptyText(spec);
+          myTargetSpecsBox.addItem(spec, spec, true);
+
+          myLibsWithSourcesCheckBox.setEnabled(true);
+        } else {
+          errors.add(String.format("Fail to find relative path from %s to build root.", file.getPath()));
+          return;
+        }
+        break;
+
+      case executable:
+        myTargetSpecsBox.setEnabled(false);
+        myTargetSpecsBox.setEmptyText(PantsUtil.getRelativeProjectPath(file.getPath()).orElse(file.getName()));
+
+        myLibsWithSourcesCheckBox.setSelected(false);
+        myLibsWithSourcesCheckBox.setEnabled(false);
+        break;
+
+      case isBUILDFile:
+        myTargetSpecsBox.setEnabled(true);
+        myTargetSpecsBox.setEmptyText(StatusText.DEFAULT_EMPTY_TEXT);
+        myLibsWithSourcesCheckBox.setEnabled(true);
+
+        ProgressManager.getInstance().run(new Task.Modal(getProject(),
+            PantsBundle.message("pants.getting.target.list"), false) {
+          @Override
+          public void run(@NotNull ProgressIndicator indicator) {
+            try {
+              final Collection<String> targets = PantsUtil.listAllTargets(projectPath);
+              UIUtil.invokeLaterIfNeeded(() -> {
                 targets.forEach(s -> myTargetSpecsBox.addItem(s, s, false));
               });
-          } catch (RuntimeException e) {
-            UIUtil.invokeLaterIfNeeded((Runnable) () -> {
+            } catch (RuntimeException e) {
+              UIUtil.invokeLaterIfNeeded((Runnable) () -> {
                 Messages.showErrorDialog(getProject(), e.getMessage(), "Pants Failure");
                 Messages.createMessageDialogRemover(getProject()).run();
               });
+            }
           }
-        }
-      });
+        });
+        break;
+      default:
+        throw new ConfigurationException("unexpected state: " + pathFileType);
```

Comment:
The compile error suggests this function cannot throw `ConfigurationException`, so using `UIUtil.invokeLaterIfNeeded` like above might be more appropriate.

Reply:
last minute adjustment fail 🤦‍♂️

Reply:
Does `UIUtil.invokeLaterIfNeeded` allow the exception to bubble up somewhere visible?

Reply:
Yes the example above has `Pants Failure: ...`, which shows up as 
![screen shot 2018-07-26 at 1 47 50 pm](https://user-images.githubusercontent.com/596358/43287697-a3111d60-90da-11e8-87d3-63a6323bcb66.png)

You can test it by changing pants script to return false
```
$ cat pants
false
```